### PR TITLE
ddl: enhance row size estimation before adding index

### LINKED ISSUE #52085: Add index row estimation is not fully usable
## Enhancement

According to the add-index test result, wider tables are more likely to encounter bottlenecks at the table reader, and CPU utilization could be  very low. #52000 adopts a strategy to determine reader concurrency: it used a statistic cache to estimate average row size of a table, and adjust reader count based on it.

However, statistic cache is not always available. For example, if a table has never executed ANALYZE, there is no data in statistic cache. For large tables, even if we analyze them with a low sample rate, it is still time-consuming.

We should investigate another way to estimate average row size efficiently.

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #52085

Problem Summary: See #52085.

### What changed and how does it work?

- Move avarage row size estimation to the beginning of adding index.
- Use `approximate-key-size` * `MB` / `approximate-keys` from region info fetched by PD client to enhance row size estimation.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/ddl/backfilling_scheduler.go [hunk 1]
```diff
@@ -595,3 +603,76 @@ func (a *taskIDAllocator) alloc() int {
...
+//  1. statistic cache (require analyze table).
+//  2. region info's approximate key size / key number.
+func estimateRowSize(ctx context.Context, store kv.Storage, tbl table.Table) (sizeInBytes int) {
+	defer util.Recover(metrics.LabelDDL, "estimateRowSize", nil, false)
+	var gErr error
+	defer func() {
+		logutil.Logger(ctx).Info("estimate row size",
+			zap.Int64("tableID", tbl.Meta().ID), zap.Int("size", sizeInBytes), zap.Error(gErr))
+	}()
+	rowCount, avgRowLen, _, _ := cache.TableRowStatsCache.EstimateDataLength(tbl.Meta())
```
Comment (ywqzzy) [inline pkg/ddl/backfilling_scheduler.go:618]: statsCache only updated when we access infoschema.Table/Columns, may be read stats from mysql.stats_meta and mysql.stats_histograms;

**Review by ywqzzy — APPROVED** [review]
Rest LGTM

File: pkg/ddl/backfilling_scheduler.go (hunk 1 above)
Comment (tangenta) [reply pkg/ddl/backfilling_scheduler.go]: Updated.

**Review by zimulala — APPROVED** [review]

**Pull request merged**